lib/commands/device/move.ts

```diff
@@ -145,42 +152,56 @@ export default class DeviceMoveCmd extends Command {
 		balena: BalenaSDK,
 		devices: ExtendedDevice[],
 	) {
-		const [deviceDeviceTypes, deviceTypes] = await Promise.all([
-			Promise.all(
-				devices.map((device) =>
-					balena.models.device.getManifestBySlug(
-						device.is_of__device_type[0].slug,
-					),
+		const { getExpandedProp } = await import('../../utils/pine');
+		// deduplicate the slugs
+		const deviceCpuArchs = Array.from(
+			new Set(
+				devices.map(
+					(d) => d.is_of__device_type[0].is_of__cpu_architecture[0].slug,
 				),
 			),
-			balena.models.config.getDeviceTypes(),
-		]);
-
-		const compatibleDeviceTypes = deviceTypes.filter((dt) =>
-			deviceDeviceTypes.every(
-				(deviceDeviceType) =>
-					balena.models.os.isArchitectureCompatibleWith(
-						deviceDeviceType.arch,
-						dt.arch,
-					) &&
-					!!dt.isDependent === !!deviceDeviceType.isDependent &&
-					dt.state !== 'DISCONTINUED',
-			),
+		);
+
+		const deviceTypeOptions = {
+			$select: 'slug',
+			$expand: {
+				is_of__cpu_architecture: {
+					$select: 'slug',
+				},
+			},
+		} as const;
+		const deviceTypes = (await balena.models.deviceType.getAllSupported(
+			deviceTypeOptions,
+		)) as Array<PineTypedResult<DeviceType, typeof deviceTypeOptions>>;
+
+		const compatibleDeviceTypeSlugs = new Set(
+			deviceTypes
+				.filter((deviceType) => {
+					const deviceTypeArch = getExpandedProp(
+						deviceType.is_of__cpu_architecture,
+						'slug',
+					)!;
+					return deviceCpuArchs.every((deviceCpuArch) =>
+						balena.models.os.isArchitectureCompatibleWith(
+							deviceCpuArch,
+							deviceTypeArch,
+						),
+					);
+				})
+				.map((deviceType) => deviceType.slug),
 		);
 
 		const patterns = await import('../../utils/patterns');
 		try {
 			const application = await patterns.selectApplication(
 				(app) =>
-					compatibleDeviceTypes.some(
-						(dt) => dt.slug === app.is_for__device_type[0].slug,
-					) &&
+					compatibleDeviceTypeSlugs.has(app.is_for__device_type[0].slug) &&
 					devices.some((device) => device.application_name !== app.app_name),
 				true,
 			);
 			return application;
 		} catch (err) {
-			if (deviceDeviceTypes.length) {
```

Comment:
This was always having a length that was equal with the No of devices passed, so I guess it wasn't working at all before.